sessionctx/variable/session.go

```diff
@@ -643,7 +645,19 @@ func (s *SessionVars) SetSystemVar(name string, val string) error {
 	case TiDBCurrentTS, TiDBConfig:
 		return ErrReadOnly
 	case TiDBMaxChunkSize:
-		s.MaxChunkSize = tidbOptPositiveInt32(val, DefMaxChunkSize)
+		maxChunkSize := tidbOptPositiveInt32(val, DefMaxChunkSize)
+		if s.InitChunkSize > maxChunkSize {
+			s.MaxChunkSize = mathutil.Max(s.InitChunkSize, DefMaxChunkSize)
+			s.StmtCtx.AppendWarning(errors.Errorf("Try to set TiDB MaxChunkSize to %d small than InitChunkSize %d so force use %d",
+				maxChunkSize, s.InitChunkSize, s.MaxChunkSize))
+		} else {
+			s.MaxChunkSize = maxChunkSize
+		}
+	case TiDBInitChunkSize:
+		s.InitChunkSize = tidbOptPositiveInt32(val, DefInitChunkSize)
+		if s.InitChunkSize > s.MaxChunkSize {
+			s.InitChunkSize = s.MaxChunkSize
+		}
```

Comment:
Shall we put this validation into `ValidateSetSystemVar`? Otherwise, we would only impose this check on session variables, and has no validation on global variable.

Reply:
🤔 Restrict the init chunk size to be session scope?

Reply:
- `tidb_init_chunk_size` should not be set too big
- `tidb_max_chunk_size` should not be set too small and too big

So we have two ranges:
- lowerBound4Init <= tidb_init_chunk_size <= upperBound4Init
- lowerBound4Max <= tidb_max_chunk_size <= uppperBound4Max

If upperBound4Init <= lowerBound4Max is guaranteed, we don't need to compare the two values when set session or global chunk sizes.

How about set `upperBound4Init` and `lowerBound4Max` to 1024? If we want to test the behavior of multi-Chunk result, we just need to set `tidb_init_chunk_size` to 1.

Reply:
It's hard to maintain the constraints between multiply concurrent sessions. We may need a lock for both values, a simplified routine is:

```
mu.Lock()
read v1
compare v1 and v2
set v2 according to the compare result
mu.Unlock()
```

Reply:
I had try to implement it with tidb's `for update` + self-retry way in https://github.com/pingcap/tidb/pull/8480/files#diff-928cb614ae092ed4c6f349f4c4c991e4R874

Reply:
I think we don't need this complicated behavior. As I mentioned here https://github.com/pingcap/tidb/pull/8480/files#r245004446, we just need to make sure the valid range of them is not overlapped. It's simpler, and can satisfy our needs.

Reply:
it's not very complicated, and in future we maybe meet more configuration items like this need cross-field validation.

Reply:
and another question is which values should we choosen for  lowerBound4Init,upperBound4Init, lowerBound4Max,uppperBound4Max ?

Reply:
initChunkSize maybe need be same as maxChunkSize but isn't 1024, for example limitExecutor need a chunk which no need grow but with fixed N size

Reply:
> 1.it's easy to implement

easy to implement but lose real function, the easiest implement is do nothing, but we need full function instead of the easy implement.

give the implicit rule, no reason magic number limit to confuse the user
why cannot set or configure initChunkSize, maxChunkSize to same value expect 1024, why?
why cannot choose maxChunkSize limit small than 1024?
I have big memory and whole cluster generate big report result, why cannot use initChunkSize > 1024?

> 2. it's not error-prone
> 3. it's easy to maintain

why global check cannot maintain and error-prone?

Reply:
let's vote tomorrow~ :rofl: